Improve readme

emmahone opened the pull request:
Add high level overview diagram for the operator and for how users are expected to interact with the CRs managed by the operator.

jsafrane commented on README.md:
````diff
@@ -1,8 +1,35 @@
+## How does a user interact with LSO?
+```mermaid
+graph LR
+    User((User)) -->|Enables| LocalVolumeDiscovery
````
IMO, user `Creates` LocalVolumeDiscovery.
And LSO creates / manages LocalVolumeDiscoveryResult as the consequence.

emmahone replied:
Agreed. I made this change and added LVDR as well in https://github.com/openshift/local-storage-operator/commit/f8acb21ec74b51b9fdfde6ab9f8c0a8d71ad3ea9#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R25

jsafrane commented on README.md:
```diff
@@ -1,8 +1,35 @@
+    Pod -->|References a PVC| PVC
+    PV -->|References| LocalVolume
+    StorageClass -->|Creates| PV
+    VolumeAttachment -->|Correlates| Node
+    VolumeAttachment -->|Correlates| PV
```
LSO PVs are not "attachable" and do not use VolumeAttachments.

jsafrane commented on README.md:
```diff
@@ -1,8 +1,35 @@
+    LocalVolume -->|Represents a physical storage device already mounted on a node| Node
+    Pod -->|References a PVC| PVC
+    PV -->|References| LocalVolume
+    StorageClass -->|Creates| PV
```
No, LSO creates PVs. StorageClass is just for reference and is not really used anywhere in PV creation / deletion / binding.

emmahone replied:
Thanks for catching that. Changed the flow so that LSO creates the PVs in https://github.com/openshift/local-storage-operator/commit/f8acb21ec74b51b9fdfde6ab9f8c0a8d71ad3ea9

jsafrane commented on README.md:
````diff
@@ -1,8 +1,35 @@
+```mermaid
+graph LR
+    PVC -->|Requests local storage using a StorageClass| StorageClass
+    StorageClass -->|Defines a LocalVolume| LocalVolume
````
StorageClass is just for reference and is not really used anywhere in PV creation / deletion / binding. The don't define anything.

emmahone replied:
Thanks for reviewing/clarifying. This should be fixed in https://github.com/openshift/local-storage-operator/commit/f8acb21ec74b51b9fdfde6ab9f8c0a8d71ad3ea9

jsafrane commented on README.md:
```diff
@@ -1,8 +1,35 @@
+graph LR
+    PVC -->|Requests local storage using a StorageClass| StorageClass
+    StorageClass -->|Defines a LocalVolume| LocalVolume
+    LocalVolume -->|Represents a physical storage device already mounted on a node| Node
```
LocalVolume does not need to be mounted on a node, it can be just a block device.

emmahone replied:
Do you have a recommendation about what we should state as the relationship between the localvolume and the node? Right now I just have "Represents a physical storage device already mounted/existent on a node".

jsafrane commented on README.md:
```diff
@@ -1,8 +1,35 @@
+    LSO((Local Storage Operator)) -->|Manages| LocalVolume
+    LSO((Local Storage Operator)) -->|Manages| StorageClass
+    LocalVolumeSet -->|Discovers local storage devices on node| Node
+    LocalVolumeSet -->|Automatically manages LocalVolume objects|LocalVolume
```
LocalVolumeSet does not manages LocalVolumes, it directly creates PVs.

emmahone replied:
Changed flow so that LVS creates PVs https://github.com/openshift/local-storage-operator/commit/f8acb21ec74b51b9fdfde6ab9f8c0a8d71ad3ea9#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R13

emmahone commented:
Thanks so much for the review @jsafrane ! I think I have addressed most of those points you made. Is there a best practice for contributing to LSO? Should I squash all these changes into a single commit?

gnufied commented on README.md:
```diff
@@ -1,8 +1,31 @@
+    User -->|Creates| LocalVolume
+    LocalStorageOperator((LocalStorageOperator)) -->|Creates/Manages| LocalVolumeDiscoveryResult
+    LocalStorageOperator -->|Manages| LocalVolume
+    LocalStorageOperator -->|Manages| StorageClass
```
If we are going to say local-storage-operator manages `LocalVolume` objects then it also manages `LocalVolumeSet` objects. But then may be neither of them is true, because LSO only reconciles these objects. It is user which creates and deletes `LocalVolume` and `LocalVolumeSet` objects.
